bohr/snorkel_utils.py

```diff
@@ -1,20 +1,334 @@
-from typing import Optional
+from functools import wraps, cached_property, lru_cache
+from typing import Optional, List, Set, Mapping, Any, Tuple, Callable, Union
+from snorkel.types import DataPoint
+
+from bohr import TEST_DIR, TRAIN_DIR
+
+from dataclasses import dataclass, field
+from cachetools import LRUCache
+
+import pandas as pd
 
 import snorkel
-from snorkel.labeling import labeling_function
+from snorkel.labeling import labeling_function, LabelingFunction
+from snorkel.preprocess import preprocessor, BasePreprocessor
+from snorkel.map import BaseMapper, LambdaMapper
+from snorkel.map.core import MapFunction
+
+from nltk.tokenize import word_tokenize
+from nltk.stem import PorterStemmer 
+from nltk import bigrams
 
 Label = int
 
 BUG = 1
 BUGLESS = 0
 ABSTAIN = -1
 
+LABEL_NAMES = {
+    BUG: 'bug',
+    BUGLESS: 'bugless',
+}
+
+def safe_word_tokenize(text: Any) -> Set[str]:
+    if text is None: return set()
+    if pd.isna(text): return set()
+
+    return word_tokenize(str(text).lower())
+
+
+@dataclass
+class Issue:
+    title: str
+    body: str
+    labels: List[str]
+
+    @cached_property
+    def stemmed_labels(self) -> Set[str]:
+        stemmer = PorterStemmer()
+        return set([stemmer.stem(l) for l in self.labels])
+
+    @cached_property
+    def tokens(self) -> Set[str]:
+        if self.body is None: return set()
+        return safe_word_tokenize(self.body)
+
+    @cached_property
+    def ordered_stems(self) -> List[str]:
+        stemmer = PorterStemmer()
+        return [stemmer.stem(w) for w in self.tokens]
+
+    @cached_property
+    def stems(self) -> Set[str]:
+        return set(self.ordered_stems)
+
+    @cached_property
+    def stem_bigrams(self) -> Set[Tuple[str, str]]:
+        return set(bigrams(self.ordered_stems))
+
+
+
+class Issues:
+    def __init__(self, issues):
+        self.__issues = issues
+
+    def __len__(self) -> int:
+        return len(self.__issues)
+
+    def __getitem__(self, idx) -> Issue:
+        return self.__issues[idx]
+
+    def match_label(self, stemmed_labels: Set[str]) -> bool:
+        for issue in self.__issues:
+            if not issue.stemmed_labels.isdisjoint(stemmed_labels): return True
+        return False
+
+    def contains_label(self, stemmed_label: str) -> bool:
+        for issue in self.__issues:
+            if stemmed_label in issue.stemmed_labels: return True
+        return False
+
+    def match(self, stemmed_keywords: Set[str]) -> bool:
+        for issue in self.__issues:
+            if not issue.stems.isdisjoint(stemmed_keywords): return True
+        return False
+
+    def contains(self, stemmed_keyword: str) -> bool:
+        for issue in self.__issues:
+            if stemmed_keyword in issue.stems: return True
+        return False
+
+    def match_bigram(self, stemmed_bigrams: Set[Tuple[str, str]]) -> bool:
+        for issue in self.__issues:
+            if not issue.stem_bigrams.isdisjoint(stemmed_bigrams): return True
```

Comment:
sometimes `stemmed_bigrams` ends up being None, and then an error is thrown.

Possible fix: https://github.com/giganticode/bohr/blob/cce5071b5ea11b25bddcc6dd444c2d58a699b80d/bohr/snorkel_utils.py#L278, the default value for `bigram` param should be `set()` ?